src/fu-bluez-backend.c

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright (C) 2021 Ricardo Cañuelo <ricardo.canuelo@collabora.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#define G_LOG_DOMAIN				"FuBackend"
+
+#include "config.h"
+
+#include "fu-bluez-device.h"
+#include "fu-bluez-backend.h"
+
+struct _FuBluezBackend {
+	FuBackend		 parent_instance;
+	GDBusConnection		*connection;
+	GHashTable		*devices;	/* platform_id : * FuDevice */
+};
+
+G_DEFINE_TYPE (FuBluezBackend, fu_bluez_backend, FU_TYPE_BACKEND)
+
+/*
+ * Returns a new FuBleDevice populated with the properties passed
+ * in the properties variant (@a{sv}).
+ *
+ * TODO: Search device characteristics and populate the hash table.
+ */
+static FuBluezDevice *
+fu_bluez_load_device_properties (GVariant *properties)
+{
+	const gchar *prop_name;
+	GVariantIter it;
+	g_autoptr(FuBluezDevice) dev = fu_bluez_device_new ();
+	g_autoptr(GVariant) prop_val = NULL;
+
+	g_variant_iter_init (&it, properties);
+	while (g_variant_iter_next (&it, "{&sv}", &prop_name, &prop_val)) {
+		if (g_getenv ("FU_BLUEZ_BACKEND_DEBUG") != NULL) {
+			g_autofree gchar *str = g_variant_print (prop_val, TRUE);
+			g_debug ("%s: %s", prop_name, str);
+		}
+		if (g_strcmp0 (prop_name, "Address") == 0) {
+			fu_ble_device_set_address (FU_BLE_DEVICE (dev),
+						   g_variant_get_string (prop_val, NULL));
+			continue;
+		}
+		if (g_strcmp0 (prop_name, "Adapter") == 0) {
+			fu_ble_device_set_adapter (FU_BLE_DEVICE (dev),
+						   g_variant_get_string (prop_val, NULL));
+			continue;
+		}
+		if (g_strcmp0 (prop_name, "Name") == 0) {
+			fu_device_set_name (FU_DEVICE (dev),
+					    g_variant_get_string (prop_val, NULL));
+			continue;
+		}
+		if (g_strcmp0 (prop_name, "Icon") == 0) {
+			fu_device_add_icon (FU_DEVICE (dev),
+					    g_variant_get_string (prop_val, NULL));
+			continue;
+		}
+		if (g_strcmp0 (prop_name, "Modalias") == 0) {
+			fu_bluez_device_set_modalias (dev, g_variant_get_string (prop_val, NULL));
+			continue;
+		}
+		if (g_strcmp0 (prop_name, "Connected") == 0) {
+			if (g_variant_get_boolean (prop_val))
+				fu_device_add_flag (FU_DEVICE (dev), FWUPD_DEVICE_FLAG_CONNECTED);
+			continue;
+		}
+	}
```

Comment:
How do we avoid a situation that the device is connected but disconnects (out of range for example)?

Reply:
I think we'll have to create a proxy for this (for each device?) and then *set* the properties based on a requery. I'll leave this to @hardboprobot tho, I'm nearly out of time today.

Reply:
So with hardwired devices we will get events that they're plugged/unplugged and when you run `fwupdmgr get-devices` (or a simlar dbus call) you will only see the devices presently connected to the system.  There is no messaging or concept of a "saved" device.  There was an idea posted a while back https://github.com/fwupd/fwupd/discussions/1418 which talked about remembering unconnected USB devices.

If the device gets disconnected mid-update then the USB stack gets upset and an error bubbles up.

Reply:
> not relying so much on synchronous calls

At the moment I think that's fine. If we're blocking for multiple seconds frequently we can do something more clever, but don't stress about the sync-ness at the moment.

Reply:
>fu_bluez_backend_coldplug

Since signals aren't yet there - is there also a dbus signal that we'll get when a device is **paired** after coldplug?  I think that's the most common way that devices will work that I am worried about.  System is already booted up(`bluetoothd` and `fwupd` running) and then they go into the UX to pair something.  `bluetoothd` knows about it, but `fwupd` needs to learn about it after coldplug was already done.

Reply:
> but fwupd needs to learn about it after coldplug was already done

I think we need to connect to all GDBusProxy's of all objects. I think it only makes sense to `add` on connected and `remove` on disconnected, otherwise we have a lot more state to worry about.

Reply:
OK - If that's the case, then I don't think there is any reason to keep a connected attribute in the first place.

Reply:
> You can listen for connection and disconnection events

I don't think _connected_ is enough. I think it needs to be _paired_ too. I think GDBusObjectManager helps you with that. I can have a play too if that helps.

Reply:
@hardboprobot I've converted FuBluezBackend to use GDBusObjectManager -- I think it makes a lot more sense -- and also means we can match on devices that are connected **and** paired. I'll remove the new device flag now too.